### PR TITLE
Better IDE support for custom tokens

### PR DESCRIPTION
**Pull request by gharlan**

### CODE REVIEW

File: Symfony/CS/Tokenizer/AbstractTransformer.php [hunk 1]
```diff
@@ -55,8 +55,18 @@ public function registerCustomTokens()
...
     }
+
+    /**
+     * Next generated value for custom token.
+     *
+     * @return int
+     */
+    final protected static function getNextGeneratedCustomTokenValue()
+    {
+        return ++self::$lastGeneratedCustomTokenValue;
```
Comment (keradus) [inline Symfony/CS/Tokenizer/AbstractTransformer.php]: overcomplexity + performance loss?

File: Symfony/CS/Tokenizer/Transformer/ArrayTypehint.php [hunk 2]
```diff
@@ -43,4 +43,12 @@ public function getCustomTokenNames()
     {
         return array('CT_ARRAY_TYPEHINT');
     }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function registerCustomTokens()
+    {
+        define('CT_ARRAY_TYPEHINT', self::getNextGeneratedCustomTokenValue());
```
Comment (keradus) [inline Symfony/CS/Tokenizer/Transformer/ArrayTypehint.php]: in every Transformer you override method from `AbstractTransformer` to... do _the same_ thing.

**Comment by keradus** [comment]
:-1:

File: Symfony/CS/Tokenizer/Transformer/ArrayTypehint.php (hunk 2 above)
Comment (gharlan) [inline Symfony/CS/Tokenizer/Transformer/ArrayTypehint.php]: In 2.x I would remove the method from `AbstractTransformer`.

Comment (keradus) [inline Symfony/CS/Tokenizer/Transformer/ArrayTypehint.php]: keep sth simple and generic in one place or in 10... i prefer 1.

**Comment by gharlan** [comment]
I'm not sure about this, too. Just an idea.

I don't like this:

bildschirmfoto 2015-12-27 um 12 00 57

I want to get auto completion etc.

Another idea: add a `Symfony/CS/Resources/custom_tokens` with content like this:

```php
define('CT_ARRAY_TYPEHINT', false);
define('CT_CLASS_CONSTANT', false);
define('CT_CURLY_CLOSE', false);
// ...
```

So the file would not be included, it would exist for ide support only.

File: Symfony/CS/Tokenizer/Transformer/ArrayTypehint.php (hunk 2 above)
Comment (gharlan) [inline Symfony/CS/Tokenizer/Transformer/ArrayTypehint.php]: I prefer a solution that is simple to use. And in my opinion the "undefined constants" are confusing.

Comment (gharlan) [inline Symfony/CS/Tokenizer/Transformer/ArrayTypehint.php]: Of course, there should not be a performance loss, and I had to check this, if this pr had a chance to be merged.
But it seems that this is not the case.. ;)

**Comment by keradus** [comment]
> And in my opinion the "undefined constants" are confusing.

this is problem o your IDE, not of PHP CS Fixer.

**Comment by GrahamCampbell** [comment]
:-1:

**Comment by GrahamCampbell** [comment]
Never optimize code for an IDE.

**Comment by gharlan** [comment]
Ok, then closing ;)

**Pull request closed without merging**